### DISCUSSION TITLE
Libvici api blocked

### DISCUSSION THREAD

**Discussion by IgorGonLer** [strongswan/strongswan#2425] [Q&A]
Hello everyone,

I'm currently working with VPP version 23.10 on Ubuntu 22.04, alongside Strongswan 5.9.6. I'm trying to establish multiple connections using libvici within separate threads in C++. Each thread has its own vici connection. However, when I attempt to list the SAs, the application sometimes becomes unresponsive and if I check with gdb the code is blocked in the line vici_submit.
` if (vici_register(conn, cCommand, listCb, nullptr) == 0)
    {
        strcpy(cArgument, "list-sas");
        req = vici_begin(cArgument);
        res = vici_submit(req, conn);
        if (res) {
  `

Even running the command swanctl --list-sa fails to return any output.

Does anyone have any suggestions, insights, or examples that might help resolve this issue?

**Comment by tobiasbrunner** [comment]
I'd suggest you try a newer release.

**Comment by IgorGonLer** [comment]
Thank you, Tobias. I will take care of it.

By the way, I reviewed and debugged the code. I believe the issue is occurring at this point:
```
int vici_register(vici_conn_t *conn, char *name, vici_event_cb_t cb, void *user)
{
------------------
	conn->mutex->lock(conn->mutex);
	while (conn->wait == WAIT_IDLE)
	{
		conn->cond->wait(conn->cond, conn->mutex);
	}
	-------------------
```
Perhaps with this approach, the code won’t be blocked. I’m using the timed_wait function from condvar.h.
```
int vici_register(vici_conn_t *conn, char *name, vici_event_cb_t cb, void *user)
{
--------------------
    unsigned int timeout = 5000;
    conn->mutex->lock(conn->mutex);
    while (conn->wait == WAIT_IDLE)
    {
        bool timed_out = conn->cond->timed_wait(conn->cond, conn->mutex, timeout);

        if (timed_out) {
            errno = ETIMEDOUT;
            conn->mutex->unlock(conn->mutex);
            return 1;  // Timeout occurred, returning with error
        }
    }
   ------------------------- 
```
Thank you in advance,
Igor

**Reply by tobiasbrunner** [reply]
Note that you should probably not use libvici directly. It's an internal library. For external use, consider using [davici](https://github.com/strongswan/davici) instead, which also has a less restrictive license.
